### ISSUE TITLE
Using 'app' as a sub domain breaks authentication cookie persistence

### ISSUE THREAD

**Issue by chrisaliotta** [aspnet/AspNetCore#11857]
# Using 'app' as a sub domain breaks authentication  cookie persistence

We found in both our production and test environments that using 'app.yourdomain.com' breaks authentication cookie persistence (especially when the application pool resets). When testing with other sub domain names like 'test.youdomain.com' authentication cookie persistence worked even after application pool reset.

# General

DotNetCore Version:  2.2.5
Operating System: Windows Server 2019 (Version 1809)
Web Hosting: IIS v10.0.17763.1
Application Pool Settings: 
 - .NET CLR Version: No Managed Code
 - Managed Pipeline Mode: Integrated

# Steps to reproduce:

## Scenario 1
1. Create dotnetcore web api application. In startup configuration specify that authentication expires some distance in the future measured in days.

```csharp
services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
    .AddCookie(options => {
        options.Events.OnRedirectToAccessDenied = ReplaceRedirector(HttpStatusCode.Forbidden, options.Events.OnRedirectToAccessDenied);
        options.Events.OnRedirectToLogin = ReplaceRedirector(HttpStatusCode.Unauthorized, options.Events.OnRedirectToLogin);
        options.Cookie.HttpOnly = true;
        options.Cookie.Expiration = TimeSpan.FromDays(150);
        options.Cookie.SameSite = SameSiteMode.None;
        // If the LoginPath isn't set, ASP.NET Core defaults 
        // the path to /Account/Login.
        options.LoginPath = PathString.Empty;//"/account/login";
        // If the AccessDeniedPath isn't set, ASP.NET Core defaults 
        // the path to /Account/AccessDenied.
        options.AccessDeniedPath = PathString.Empty;
        options.SlidingExpiration = true;
    });
```
2. In IIS create a website with "app" as it's sub-domain. (eg. app.contoso.com)
3. Execute a login on your web api with postman.
4. Confirm that the user is authenticated by accessing a protected api end-point.
5. Go to IIS stop and restart the site.
6. Try accessing the protected api end-point,
7. You should experience a not authenticated error and the authentication cookie will no longer work.

## Scenario 2
1. Create a new website in IIS with a random sub-domain name (eg. test1.constoso.com) and copy the contents of the original site to the new location.
2. Repeat the steps 3 through 6 in scenario 1. 
3. You should still be able to access the protected web api end-point and the authentication cookie will still work.

### Expected behavior
Scenario 1 should behave the same as scenario 2.

**Comment by chrisaliotta** [comment]
Side note: I moved this issue from the dotnet/core project issues. Don't be confused by the reference above, this issue is still _very much_ open.

**Comment by blowdart** [comment]
That's odd. What errors are you seeing in the server side logs?

**Comment by chrisaliotta** [comment]
That's whats perplexing, is that we're not seeing anything abnormal. I've hunted through the Event Viewer and even looked at IIS text based log files -- nothing unusual or any errors. Are there any specific areas or events you would like us to hunt for?

**Comment by blowdart** [comment]
The asp.net core [logs](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-2.2). You have to add Windows Event log as an option (persisting to text might be easier to use)

Does this happen only under IIS? Rather than kestrel directly? A failed auth should have log entries.

**Comment by chrisaliotta** [comment]
Sounds good -- we're going to recompile with the logging features enabled and replicate the tests. I'll post the log files once we've completed the test work 👍 

Currently, we've only been able to test this with IIS so I am not sure if this affect Kestrel.

**Comment by Tratcher** [comment]
Side note: do not use `options.Cookie.Expiration = TimeSpan.FromDays(150);`, use `options.ExpireTimeSpan = TimeSpan.FromDays(150);` instead.
https://github.com/aspnet/AspNetCore/blob/33eb8ba552c7452f441e2a223457d39f271d01b7/src/Security/Authentication/Cookies/src/CookieAuthenticationOptions.cs#L62
Using options.Cookie.Expiration will throw in 3.0.

**Comment by Tratcher** [comment]
This smells like https://github.com/aspnet/DataProtection/issues/102. The app name wouldn't be special in that case, except if that app previously existed?

**Comment by blowdart** [comment]
Well, I had thought that, but the issue says IIS has been started in both the working, and the not working cases, which would rule out ephemeral keys. But that's why I want logs :)

**Comment by chrisaliotta** [comment]
Hey guys, based on our tests these were the log events we found:

```
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler
Cookies was not authenticated. Failure message: Unprotect ticket failed

Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager
No XML encryptor configured. Key {e2f099b7-dbe0-4132-ba30-43cc3f7c984a} may be persisted to storage in unencrypted form.

Microsoft.AspNetCore.DataProtection.Repositories.EphemeralXmlRepository
Using an in-memory repository. Keys will not be persisted to storage.

Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager
Neither user profile nor HKLM registry available. Using an ephemeral key repository. Protected data will be unavailable when application exits.
```

**Comment by Tratcher** [comment]
Ah, check https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/index?view=aspnetcore-2.2#create-a-data-protection-registry-hive

**Comment by chrisaliotta** [comment]
@Tratcher Thanks! That worked! We fixed the problem by configuring the IIS Application Pool to load the user profile as recommended in the link you posted. It's just strange that this issue manifested itself when we used "app" in the sub-domain name. I guess that's the only thing that bothers me about this issue :)

@blowdart Thank you for taking this issue seriously and helping us! We really appreciate you guys and what you do for the dotnet community!
